tensorflow_addons/layers/snake.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Implements Snake layer."""
+
+import tensorflow as tf
+from typeguard import typechecked
+
+from tensorflow_addons.activations.snake import snake
+
+from tensorflow_addons.utils import types
+
+
+@tf.keras.utils.register_keras_serializable(package="Addons")
+class Snake(tf.keras.layers.Layer):
+    """Snake layer to learn periodic functions with the 'frequency' param trainable.
+
+    https://arxiv.org/abs/2006.08195
+    """
+
+    @typechecked
+    def __init__(self, frequency_initializer: types.Initializer = "ones", **kwargs):
+        """
+        frequency_initializer: Initializer for the 'frequency' param.
+        """
```

Comment:
Sorry for the confusion, after a rough look at our docs, it would be better to move this near `class Snake`.

```python
class Snake(tf.keras.layers.Layer):
    """Snake layer to learn periodic functions with the trainable `frequency` scalar.

    It follows:
    ```
      f(x) = x + (1 - cos(2 * frequency * x)) / (2 * frequency)
    ```

    where `frequency` is a learned scalar.

    See [Neural Networks Fail to Learn Periodic Functions and How to Fix It](https://arxiv.org/abs/2006.08195) for more details.

    Arguments:
        frequency_initializer: Initializer for the `frequency` scalar.
        frequency_regularizer: Regularizer for the `frequency` scalar.
        frequency_constraint: Constraint for the `frequency` scalar.
    """
```

I'm following doc of the parametric activation like [PReLU](https://www.tensorflow.org/api_docs/python/tf/keras/layers/PReLU) here. Feel free to elaborate more.

Reply:
> IMO, it suggests `frequency` to be with the same shape as the input (like PReLU). What do you think?

As I said earlier, the paper is quite obscure in many aspects. The same statement could have been made about the activation. Guesswork is a Sisyphus work. So I propose to leave it as it is. We can always get back to it later.

> Also, following the layers with learnable parameter in core TF, we should also support some arguments like `frequency_regularizer` and `frequency_constraint`.

Same here. I thought of it but the paper does not give a glue on how these 'regularizers and co' should be applied. If they are relevant here at all.

Reply:
Got it :-) Then just move the doc near `class Snake`.

Reply:
I had to eliminate the math formula because it looks very awkward in the plain text. In the end of the all, there is the academic paper an one must be in good knowledge before utilizing the layer.

Reply:
ok